rescript/get_unite.py

```diff
@@ -0,0 +1,193 @@
+        },
+        "eukaryotes": {
+            False: "10.15156/BIO/2938081",
+            True: "10.15156/BIO/2938082",
+        },
+    },
+    # Old version 9.0 is not listed here
+    "8.3": {
+        "fungi": {
+            False: "10.15156/BIO/1264708",
+            True: "10.15156/BIO/1264763",
+        },
+        "eukaryotes": {
+            False: "10.15156/BIO/1264819",
+            True: "10.15156/BIO/1264861",
+        },
+    },
+    "8.2": {
+        "fungi": {
+            False: "10.15156/BIO/786385",
+            True: "10.15156/BIO/786387",
+        },
+        "eukaryotes": {
+            False: "10.15156/BIO/786386",
+            True: "10.15156/BIO/786388",
+        },
+    },
+}
+
+
+def _unite_get_url(
+    version: str = None, taxon_group: str = None, singletons: bool = None
+) -> str:
+    """Get DOI from included list, then query plutof API for UNITE url"""
+    # Get matching DOI
+    doi = UNITE_DOIS[version][taxon_group][singletons]
+    # Build URL
+    base_url = (
+        "https://api.plutof.ut.ee/v1/public/dois/"
+        "?format=vnd.api%2Bjson&identifier="
+    )
+    query_data = requests.get(base_url + doi).json()
+    # Updates can be made to files in a DOI, so on the advice of the devs,
+    # only return the last (newest) file with this -1  vv
+    URL = query_data["data"][0]["attributes"]["media"][-1]["url"]
+    return URL
+
+
+# Make tmp_dir for standalone testing
+# tmp_dir = tempfile.mkdtemp()
+# print(tmp_dir)
+
+
+def _unite_get_tgz(
+    url: str = None, download_path: str = None, retries: int = 10
+) -> str:
+    """Download compressed database"""
+    for retry in range(retries):
+        # Track downloaded size
+        file_size = 0
+        try:
+            response = requests.get(url, stream=True)
+            # Save .tgz file
+            unite_file_path = os.path.join(download_path, "unitefile.tar.gz")
+            with open(unite_file_path, "wb") as f:
+                for chunk in response.iter_content(chunk_size=8192):
+                    if chunk:
+                        f.write(chunk)
+                        file_size += len(chunk)
+            # Check if the downloaded size matches the expected size
+            if file_size == int(response.headers.get("content-length", 0)):
+                return unite_file_path  # done!
+            else:
+                raise ValueError("File download incomplete")
+        except HTTPError as e:
+            print(
+                "Request failed with code "
+                + str(e.response.status_code)
+                + ", on try "
+                + str(retry)
+            )
+        except ValueError:
+            print("File incomplete, on try " + str(retry))
+
+
+# Test it by downloading this file
+# _unite_get_tgz(_unite_get_url("8.3", "fungi", False), tmp_dir)
+# _unite_get_tgz(_unite_get_url("9.0", "fungi", False), tmp_dir)
+
+
+def _unite_get_artifacts(
+    tgz_file: str = None, cluster_id: str = None
+) -> (None, None):
```

Comment:
What is the object type of Qiime2 artifacts?

In `ncbi.py`, `get_ncbi_data()` returns a `(DNAIterator, DataFrame)` but I'm not sure that's right here...

Reply:
How do I know for sure? I've run it with both `(DataFrame, DNAFASTAFormat)` and `(Series, DNAFASTAFormat)` and it seems to not care.

The type is of the Q2 artifact
```python
>>> tmpa = _unite_get_artifacts(...)
>>> tmpa[0][0].type
FeatureData[Taxonomy]
```

Reply:
When I run this from the CLI, I get this error:

```
Traceback (most recent call last):
  File "/home/biouser/miniconda3/envs/qiime2-2023.7/lib/python3.8/site-packages/q2cli/commands.py", line 478, in __call__      
    results = self._execute_action(
  File "/home/biouser/miniconda3/envs/qiime2-2023.7/lib/python3.8/site-packages/q2cli/commands.py", line 539, in _execute_action
    results = action(**arguments)
  File "<decorator-gen-794>", line 2, in get_unite_data
  File "/home/biouser/miniconda3/envs/qiime2-2023.7/lib/python3.8/site-packages/qiime2/sdk/action.py", line 342, in bound_callable
    outputs = self._callable_executor_(
  File "/home/biouser/miniconda3/envs/qiime2-2023.7/lib/python3.8/site-packages/qiime2/sdk/action.py", line 583, in _callable_executor_
    self.signature.coerce_given_outputs(output_views, output_types,
  File "/home/biouser/miniconda3/envs/qiime2-2023.7/lib/python3.8/site-packages/qiime2/core/type/signature.py", line 485, in coerce_given_outputs
    (spec.view_type.__name__, type(output_view).__name__))
AttributeError: 'NoneType' object has no attribute '__name__'
```

Does this have to do with these `None` types?

Reply:
All tests pass. This is the last thing I need to resolve, then the CLI works!
@nbokulich, any ideas?

Reply:
It depends on how you want to make use of the data. This is where the transformers come into play. That is, there are many transformers that will help do the work of converting your data into another form. For example, you can access the taxonomy as a [DataFrame](https://github.com/qiime2/q2-types/blob/8a0e191efa5aef5a6a0d0807cf4dc9c223313036/q2_types/feature_data/_transformer.py#L164), or a [Series](https://github.com/qiime2/q2-types/blob/8a0e191efa5aef5a6a0d0807cf4dc9c223313036/q2_types/feature_data/_transformer.py#L169), etc. 

The data will be returned into whatever available format you need, and then it will be made into an artifact. Unless you are doing something to the data after you've generated it, there is no need to worry too much about if  you are returning as a DataFrame, Series, or whatever. If you look throughout many of the QIIME actions, there are ways to access taxonomy, etc.. as a variety of formats. These will be converted to the appropriate form behind the scenes. 

If you need to force a specific format, when using the API, or simply for testing purposes you can view the data a certain way, see [this](https://github.com/bokulich-lab/RESCRIPt/blob/946728467532f066c8bb750c3d8d1b8af02773b9/rescript/tests/test_derep.py#L69) for example.

Reply:
Thank you for this. I think the implicit type conversion is tripping me up

With `-> (pd.DataFrame, DNAFASTAFormat):`
I get `Expected output view type 'DataFrame', received 'Artifact'`

With `-> (qiime2.Artifact, DNAFASTAFormat):`
I get `No transformation from <class 'qiime2.sdk.result.Artifact'> to <class 'q2_types.feature_data._format.TSVTaxonomyDirectoryFormat'>`

This is custom Qiime2 stuff so I'm way out of my depth here...

Reply:
I'm thinking through how implicit type conversion works.

Maybe I should use `pd.DataFrame, DNAFASTAFormat` without calling `qiime2.Artifact.import_data()`? That matches the ncbi.py code and seems to use implicit type conversion

Reply:
My bad, I forgot to mention the syntactic sugar we often use. Replace this:

```
tax_results = qiime2.Artifact.import_data ...
...
seq_results = qiime2.Artifact.import_data ...
```
with 
```
tax_results, = qiime2.Artifact.import_data ...
...
seq_results, = qiime2.Artifact.import_data ...
```

*Note the `,`.*

As we do [here](https://github.com/bokulich-lab/RESCRIPt/blob/946728467532f066c8bb750c3d8d1b8af02773b9/rescript/get_gtdb.py#L49)

Reply:
Remember also that I'm not using a pipeline (no ctx), but I am using a function.

I should only use qiime2.Artifact.import_data in the context of a ctx pipeline?

Reply:
Sorry, that was my fault for suggesting a dataframe. I missed that you are explicitly returning a taxonomy artifact via importing. Therefor you need to return the format of your taxonomy artifact. So, I think you'd simply return (HeaderlessTSVTaxonomyFormat, DNAFASTAFormat)

Reply:
Ignore my comment (which I deleted) about the `,=` I was confusing myself. 🤦

Reply:
you need to specify appropriate view types, e.g., pandas.DataFrame or whatever makes sense... e.g., see here: https://github.com/bokulich-lab/RESCRIPt/blob/master/rescript/ncbi.py#L77

Reply:
My current code with `-> (HeaderlessTSVTaxonomyFormat, DNAFASTAFormat):`
gives this error:
`  Expected output view type 'HeaderlessTSVTaxonomyFormat', received 'Artifact'`

Reply:
correct — you should not import these as artifacts (assuming that you are writing a `method`, not a `pipeline`)

Reply:
Okay, I am a dunce. I forgot that there is quite a difference in how things are handled via a pipeline and just a simple method. Pipelines can deal directly with artifacts, and pass them down to other steps. The methods, do not. So, to set it up as *not* a pipeline do the following:

In the `_unite_get_artifacts` function:
```
tax_results = TaxonomyFormat(fp, mode='r').view(DataFrame)
...
seq_results = DNAFASTAFormat(fp, mode='r').view(DNAIterator)
```

In the `get_unite_data` function do:
```
-> (DataFrame, DNAIterator)
```

Then at the top simply import:
```
from pandas import Series, DataFrame
from q2_types.feature_data import (TaxonomyFormat, DNAFASTAFormat,
                                   DNAIterator)

I just edited my version of your code like this and it worked. :-)